File: CW/Lista_3/Z3.scala
def insert[A](cmp: (A, A) => Boolean, x: A, list: List[A]): List[A] =
    list match
        case Nil => List(x)
        case y :: ys => if (cmp(x, y)) x :: list else y :: insert(cmp, x, ys)

def insertionSort[A](cmp: (A, A) => Boolean, list: List[A]): List[A] =
    list match
        case Nil => List()
        case h :: t => insert(cmp, h, insertionSort(cmp, t))

///////////////////////////////////////////////

def merge[A](cmp: (A, A) => Boolean, left: List[A], right: List[A]): List[A] =
    (left, right) match
        case (Nil, ys) => ys
        case (xs, Nil) => xs
        case (x :: xs, y :: ys) =>
            if (cmp(x, y)) x :: merge(cmp, xs, y :: ys)
            else y :: merge(cmp, x :: xs, ys)

def split[A](list: List[A]): (List[A], List[A]) =
    list match
        case Nil => (List(), List())
        case List(x) => (List(x), List())
        case x :: y :: zs =>
            val (left, right) = split(zs)
            (x :: left, y :: right)

def mergeSort[A](cmp: (A, A) => Boolean, list: List[A]): List[A] =
    list match
        case Nil => List()
        case List(x) => List(x)
        case xs =>
            val (left, right) = split(xs)
            merge(cmp, mergeSort(cmp, left), mergeSort(cmp, right))

@main def main() =
    val lst = List(5,7,4,2,56,87,123,65)
    println(mergeSort(((x: Int, y: Int) => x < y), lst))
    println(insertionSort(((x: Int, y: Int) => x < y), lst))
    println(split(List(1,2,3,4,5,6,7,8,9,10)))

File: CW/Lista_3/Z3.ml
let rec insert cmp x = function
  | [] -> [x]
  | y :: ys as list -> if cmp x y then x :: list else y :: insert cmp x ys

let rec insertionSort cmp = function
  | [] -> []
  | x :: xs -> insert cmp x (insertionSort cmp xs)

(*///////////////////////////////////////*)

let rec merge cmp left right =
  match left, right with
  | [], ys -> ys
  | xs, [] -> xs
  | x::xs, y::ys -> 
    if cmp x y then x :: merge cmp xs (y :: ys)
    else y :: merge cmp (x :: xs) ys

let rec split lst =
  match lst with
  | [] -> [], []
  | [x] -> [x], []
  | x :: y :: zs -> 
    let left, right = split zs in
    (x :: left, y :: right)

let rec mergesort cmp = function
  | [] -> []
  | [x] -> [x]
  | lst ->
    let left, right = split lst in
    merge cmp (mergesort cmp left) (mergesort cmp right)